Do not insert Related Posts feature on FSE sites

stacimc opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->
**Context**
The Related Posts feature is automatically appended to the bottom of posts on the frontend provided certain criteria are met. Particularly: the feature must be enabled (via Settings -> Reading), the blog itself must have at least 10 posts, and the post being viewed must not already contain a Related Posts block.

For FSE sites, we should not automatically append the feature here. Instead, it should be configured by the user in the site editor. This PR accomplishes this, and also fixes a few minor compatibility issues with FSE listed below.

It does **not** enqueue the assets when the block is inserted into a template part. **That change is added to the associated diff.**

Fixes #19448

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Prevents the Related Posts feature from being automatically appended to post-content for sites with an FSE theme enabled.
* Replaces the `you haven't published enough posts with similar content` preview message in placeholders with a more user friendly message when in the site editor.
* Fixes an existing bug where the Related Posts scripts are enqueued when they are not needed. These scripts are only needed for the feature added via the hook for non-AMP requests -- in this case we append client-rendered html. In all other cases the block is rendered server-side and the scripts are not only not required, but result in a console error if they are loaded.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:

At the moment, I can test this change in my sandbox by manually copying the changes over to the `mu-plugins/jetpack-plugin/vendor/automattic/jetpack-blocks/src/` directory. To do so you might use a command like: `scp ~/dev/a8c-projects/jetpack/projects/packages/blocks/src/class-blocks.php wpcom:/home/wpcom/public_html/wp-content/mu-plugins/jetpack-plugin/vendor/automattic/jetpack-blocks/src/class-blocks.php`.

**Setup**
* Create a test FSE site using the TT1 blocks theme. **To begin, it should have few fewer than 10 posts**, and the Related Posts feature must be enabled via the `Settings -> Reading` page (this should be enabled by default).
* Create a test non-FSE site using the TT1 classic theme. It should also have at least 10 posts, and the Related Posts feature should be enabled.

Verify the bugs before syncing changes:
- [ ] Open the site editor for your FSE site and insert a Related Posts block in the template. Notice the confusing placeholder text ("You haven't published enough posts"). View a post on the frontend; you should see the Related Posts, but the styles are not correct (see screenshot). Remove the block from the template before continuing.
- [ ] **Add some more posts to the FSE site now; you need at least ten.** Create a new post on the FSE site which does not contain the Related Post block. View the post on the frontend; you should see the Related Posts feature has been automatically appended to the bottom of the post. 
- [ ] Create a post on the non-FSE site and insert the Related Posts block into the content. View on the frontend, and you will see a console error ( `Uncaught TypeError: Cannot read property 'hasAttribute' of null` ) caused by the scripts being loaded unnecessarily.


Sync these changes **by applying the diff.** <- The diff contains an additional change for wpcom.

**Test the FSE site**
- [ ] View a post on the frontend again; you should now see that the Related Posts feature has **not** been automatically appended.
- [ ] Edit the single post template in the site editor and add a Related Posts block again. You should see the new placeholder message this time (see screenshots).
- [ ] View a post on the frontend; you should see the block in the footer, and this time the styling should be correct.
- [ ] Edit this post in the block editor and add a Related Posts block. You should **not** see the new placeholder message in the block editor; you should see real previews of related posts.
- [ ] View the post on the frontend and you should see _both_ Related Posts blocks.
- [ ] Turn off the feature via the `Settings -> Reading` page and verify that you do not see _any_ Related Posts blocks on the frontend.
- [ ] Turn the setting back on and verify that you can see both blocks on the frontend again.
- [ ] Remove the Related Posts blocks from both the post and from the template part via the block and site editor respectively. View the post on the frontend and you should see both blocks removed; you should also see that the Related Posts feature has _not_ been automatically appended.
- [ ] Open the console and verify that you do not see any console warnings.

**Verify that there are no regressions on your non-FSE site**
- [ ] Open a post on the frontend that does not contain a Related Posts block. You should see that the feature has automatically been appended to the bottom of the post.
- [ ] Insert a Related Posts block in the editor. View the post on the frontend and verify that you see the block where you inserted it, but that the feature was _not_ automatically appended as well. Verify there are no console errors.
- [ ] Turn off the setting via `Settings -> Reading` and verify that the block is no longer visible on the frontend.
- [ ] Without turning the setting back on, remove the block from the editor and verify that the feature still isn't automatically inserted on the frontend.
- [ ] Turn the setting back on and verify that the feature is automatically appended once again.

#### Screenshots

A Related Post block inserted into a template part on the frontend:
| Before | After |
| ---- | ---- |
| <img width="568" alt="Screen Shot 2021-04-21 at 4 58 03 PM" src="https://user-images.githubusercontent.com/63313398/115636477-e2f51600-a2c2-11eb-9faf-58169cef0fe8.png"> | <img width="527" alt="Screen Shot 2021-04-21 at 4 58 58 PM" src="https://user-images.githubusercontent.com/63313398/116158554-b74ba480-a6a3-11eb-95ac-fc3d9554592e.png"> |



A Related Post block inserted into a template part in the site editor:
| Before | After |
| ---- | ---- |
| <img width="652" alt="Screen Shot 2021-04-21 at 5 01 32 PM" src="https://user-images.githubusercontent.com/63313398/115636674-4b43f780-a2c3-11eb-8ecb-56dd063b7116.png"> | <img width="524" alt="Screen Shot 2021-04-21 at 5 01 19 PM" src="https://user-images.githubusercontent.com/63313398/115636671-497a3400-a2c3-11eb-9a1f-ae19234e70e8.png"> |

stacimc committed:
[not verified] Do not insert Related Posts feature on FSE sites

stacimc committed:
[not verified] Use Gutenberg is_fse_theme function

stacimc committed:
[not verified] Display better error message for placeholders in site editor

stacimc committed:
[not verified] Return early if related-posts block is not present on the page

Temporarily, the Related Posts assets will be enqueued for FSE sites
with the feature enabled even if the block is not present (this will
be fixed in a future iteration). To avoid console errors in this
state we should return early from the script if the block is not
found.

stacimc committed:
[not verified] Return as early as possible when block not present

stacimc committed:
[not verified] Add changelog

stacimc committed:
[not verified] Mock RelatedPosts to ensure the feature is added in test

stacimc committed:
[not verified] Move asset enqueueing to render_block

For FSE sites, it is no longer reliable to assume that assets should
be enqueued when the Related Posts feature is enabled, because the
actual block is not necessarily present on the page. This commit
moves the enqueueing of assets to the `render_block` method instead,
so that they are only enqueued when necessary. This also allows us
to remove the null checks in the scripts.

For Related Posts that are inserted via the hook on non-FSE sites,
this would break the client-rendered posts for non-AMP requests, so
we also need to enqueue assets within that branch.

stacimc committed:
[not verified] Enqueue styles but not scripts for server-rendered block

andrewserong reviewed (dismissed):
Thanks for all the work here @stacimc and the thorough testing instructions! This looks like a good clean fix to me.

✅ Confirmed the existing issues in an FSE site and a non-FSE site in wpcom
✅ Confirmed the fix in an FSE site and tested that the show/hide settings in `Settings -> Reading` behave as we'd expect (allows for a global hide of the feature)
✅ Confirmed this fixes the enqueuing issues
✅ Confirmed this doesn't introduce regressions for the non-FSE site, including the show/hide settings in `Settings -> Reading`

I like the choice for the placeholder text in the site editor!

Tested within the single post template:

![image](https://user-images.githubusercontent.com/14988353/116511555-f5e29880-a909-11eb-880a-3934541de73b.png)

And within a Footer template part:

![image](https://user-images.githubusercontent.com/14988353/116511627-101c7680-a90a-11eb-80d6-12d2c58c2055.png)

I also noticed we can (fairly) safely leave the Related Posts block in a Footer template that's used elsewhere in the site. If we do, then it'll show related posts of the same post type. So, on the Home page I'll see related pages like the About page:

![image](https://user-images.githubusercontent.com/14988353/116511929-920c9f80-a90a-11eb-9d97-d573e129e0e8.png)

But if we access a single post, it'll show related posts:

![image](https://user-images.githubusercontent.com/14988353/116511979-a51f6f80-a90a-11eb-81ac-fc85eff2b5e1.png)

I can't quite work out whether or not that's a useful feature, I just thought it was kind of neat that it still works, though I imagine mostly we'd be using the block within the single post template part.

Anyway, nice work here, this LGTM! 🚀 CI is complaining that the PR isn't up to date, so it probably just needs a `git merge origin/master` to get that action passing.

andrewserong reviewed (dismissed):
I just gave this and the wpcom diff a quick smoke test after the rebase, and it's still looking good to me! 👍

stacimc commented on projects/plugins/jetpack/modules/related-posts/jetpack-related-posts.php:
```diff
@@ -85,6 +85,15 @@ protected function get_blog_id() {
+	 * @return bool True if the current theme is an FSE/Site Editor theme
+	 */
+	protected function is_fse_theme() {
+		return function_exists( 'gutenberg_is_fse_theme' ) && gutenberg_is_fse_theme();
```
I agree! Do you think it would be reasonable to extract this in a follow-up PR? Happy to include it here if preferred, I'll just need to sort out the testing.

stacimc committed:
[not verified] Move is_fse_theme to Blocks package

andrewserong commented on projects/plugins/jetpack/modules/related-posts/jetpack-related-posts.php:
```diff
@@ -174,14 +174,15 @@ public function get_headline() {
 	 */
 	public function filter_add_target_to_dom( $content ) {
-		if ( has_block( 'jetpack/related-posts' ) ) {
+		if ( has_block( 'jetpack/related-posts' ) || Blocks::is_fse_theme() ) {
```
To avoid mid-deploy fatals, should we add a `function_exists` check before calling this method? Just thinking about when this change lands, during deployment this file could be updated before the `class-blocks.php` file is updated since that change is deployed as part of a separate package (I think!)

stacimc replied:
I believe this will be okay because of the [`use Automattic\Jetpack\Blocks;` ](https://github.com/Automattic/jetpack/pull/19615/files#diff-9105b068cf482dd9007ac0a2ff0985bf13dac7d398fb802678224448b5897841L4). (The constructor in this class is already using `jetpack_register_block` from the Blocks package.)

andrewserong replied:
I was wondering if this one might be slightly trickier. In the generated wpcom diff, it doesn't appear to include the change to `class-blocks.php` (I assume because it's part of the blocks package?). I'm not _quite_ sure the process for doing package updates, but Jeremy handled it for me in https://github.com/Automattic/jetpack/pull/19197. We might either need to manually add the change to the wpcom diff, or add the check so that the two changes can be deployed separately?

stacimc replied:
Ah sorry, I see what you mean -- since the diff already needs manual changes anyway, I've gone ahead and added the change to `class-blocks.php` to the diff as well. @jeherve Does that look like the right approach to you?

jeherve replied:
We'll actually do things in 3 parts, once that PR is merged:

1. We'll release a new version of the Blocks package. This is done by running `tools/changelogger-release.sh packages/blocks` while on `master`, and then committing the output in a new branch, and opening a PR. Once that PR gets merged, a new version of the Blocks package will automatically become available in Packagist.
2. On WordPress.com, update the version of Blocks. Here is an example from the last time I did it: D59174-code. That will bring the new method to WordPress.com simple.
3. Commit D60508-code to start using the new method.

That should work fine in this case. This turned out to be a bit more involved this time around; in the future I think it would make sense, whenever making changes to the Blocks package (or any other package), to do the changes to the package in its own PR, and then start using those changes in a separate PR, to keep PRs small.

stacimc replied:
Thanks for explaining that @jeherve -- I'll definitely break things out in the future. 

Testing my understanding: I've added the wpcom manual change back to the diff after rebasing (to prevent early asset enqueueing now that this PR has changed how that is handled), but I have _not_ added the `class-blocks.php` changes, since that will be a separate diff (your Step 2).

andrewserong commented on projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-posts.php:
```diff
@@ -882,7 +882,14 @@ function test_remove_related_posts_from_filtered_content() {
+		$jetpack_relatedposts_mock
+			->method( 'is_fse_theme' )
+			->willReturn( false );
+		$jetpack_relatedposts_mock->action_frontend_init();
```
Since `Jetpack_RelatedPosts` no longer has an `is_fse_theme` method, and the `Blocks::is_fse_theme` method can't be mocked directly because it's a static method, I wonder if we can add a filter to the `Blocks::is_fse_theme` method in a similar approach to `Blocks::is_amp_request()` and `Blocks::is_standalone_block()`. Then in the test, you can add a filter to return the value you expect? That's the approach I used in https://github.com/Automattic/jetpack/pull/19197/files in case it helps.

Though from looking at this test, I don't quite understand why it needs to be mocked, since I would have thought that `Blocks::is_fse_theme()` should be returning false by default? Unless the build in CI running the function that calls `Blocks::is_fse_theme()` doesn't include the update to the Blocks package, in which case attempting to call that function could be erroring out earlier, which could result in the related blocks content not being appended. I haven't double-checked any of that, though, so hopefully it isn't too much of a rabbit hole!

stacimc replied:
Thank you for the filter suggestion! That has it working. I also don't quite understand why it needs to be mocked, since `Blocks::is_fse_theme()` is false by default in the Blocks package tests.

andrewserong replied:
Glad that got it working! It is a little bemusing, though!

stacimc committed:
[not verified] Add filter to is_fse_theme to allow for easier testing

stacimc committed:
[not verified] Add changelog for Blocks package

jeherve committed:
Merge remote-tracking branch 'origin/master' into fix/related-posts-for-fse

jeherve committed:
Update block package version to start new cycle

jeherve committed:
Add one more test for the core function

jeherve committed:
Update version of the Blocks package used by the Jetpack plugin

jeherve commented on projects/plugins/jetpack/modules/related-posts/jetpack-related-posts.php:
```diff
@@ -174,14 +174,15 @@ public function get_headline() {
 			if ( class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ) {
 				$content .= "\n" . $this->get_server_rendered_html();
 			} else {
+				$this->_enqueue_assets( true, true );
```
It took me some time to understand the juggling that was happening when enqueuing assets. Do you think we could add some comments to make that clearer to our futureselves?

stacimc replied:
Agreed! Thank you very much for the additional test / updates 💯

jeherve reviewed (commented):
Thanks for making the changes to the Blocks package. I added an additional test, updated the PR against the latest `master`, fixed those version dependency issues. I also updated the testing instructions to take into account the changes discussed here: https://github.com/Automattic/jetpack/pull/19615#discussion_r627589945

I think it's looking good right now. 

I only have one remark below, maybe nitpicking. Let me know! :)

stacimc committed:
Add comments to explain asset enqueueing

jeherve reviewed (dismissed):
I ran some more tests, and I think the asset enqueuing changes are breaking the existing Related Posts shortcode. When I add the `[jetpack-related-posts]` shortcode to an existing post, the related posts aren't displayed when on this branch.

Do you think you could take a look?

stacimc committed:
Merge remote-tracking branch 'origin/master' into fix/related-posts-for-fse

stacimc committed:
Move asset enqueueing into client render function

Previously this was done just before calling get_client_rendered_html,
which works for related posts inserted via the hook -- but failed
to support related posts added via the shortcode, which also uses
this render method. After this change assets are enqueued only
from within their respective server-/client-side render functions,
which should also improve clarity.

samiff reviewed (approved):
Instructions test well for me 👍  Was able to use the shortcode without error as well now.

jeherve commented:
Noting that the new version of the Blocks package, including the changes from this PR, is now active on WordPress.com since r225578-wpcom. You should consequently be able to deploy the matching WordPress.com diff.

stacimc commented:
r225816-wpcom. Thanks for all your help @jeherve!
